hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/metrics/TimelineServiceV1Publisher.java

```diff
@@ -59,16 +65,79 @@ public TimelineServiceV1Publisher() {
   }
 
   private TimelineClient client;
+  private LinkedBlockingQueue<TimelineEntity> entityQueue;
+  private ExecutorService sendEventThreadPool;
+  private int dispatcherPoolSize;
+  private int dispatcherBatchSize;
+  private int putEventInterval;
+  private boolean isTimeLineServerBatchEnabled;
+  private volatile boolean stopped = false;
+  private PutEventThread putEventThread;
+  private Object sendEntityLock;
 
   @Override
   protected void serviceInit(Configuration conf) throws Exception {
+    isTimeLineServerBatchEnabled =
+        conf.getBoolean(
+            YarnConfiguration.RM_TIMELINE_SERVER_V1_PUBLISHER_BATCH_ENABLED,
+            YarnConfiguration.DEFAULT_RM_TIMELINE_SERVER_V1_PUBLISHER_BATCH_ENABLED);
+    if (isTimeLineServerBatchEnabled) {
+      putEventInterval =
+          conf.getInt(YarnConfiguration.RM_TIMELINE_SERVER_V1_PUBLISHER_INTERVAL,
+              YarnConfiguration.DEFAULT_RM_TIMELINE_SERVER_V1_PUBLISHER_INTERVAL)
+              * 1000;
+      dispatcherPoolSize = conf.getInt(
+          YarnConfiguration.RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE,
+          YarnConfiguration.
+              DEFAULT_RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE);
+      dispatcherBatchSize = conf.getInt(
+          YarnConfiguration.RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE,
+          YarnConfiguration.
+              DEFAULT_RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE);
```

Comment:
Would you validate these parameters in case they are set to negative value?

Reply:
Do a Simple check like this is fine?
```
      if (putEventInterval < 0) {
        throw new Exception("RM_TIMELINE_SERVER_V1_PUBLISHER_INTERVAL cannot be negative");
      }
      dispatcherPoolSize = conf.getInt(
          YarnConfiguration.RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE,
          YarnConfiguration.
              DEFAULT_RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE);
      if (dispatcherPoolSize < 0) {
        throw new Exception("RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE cannot be negative");
      }
      dispatcherBatchSize = conf.getInt(
          YarnConfiguration.RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE,
          YarnConfiguration.
              DEFAULT_RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE);
      if (dispatcherBatchSize < 0) {
        throw new Exception("RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE cannot be negative");
      }
```

Reply:
Thank you. I think the value should be positive.
- putEventInterval should be more than 0.
- dispatcherPoolSize should be more than 0.
- dispatcherBatchSize should be more than 1 because doing one-by-one is useless in batch.

Reply:
How does this look?
```
      if (putEventInterval < 1) {
        throw new Exception("RM_TIMELINE_SERVER_V1_PUBLISHER_INTERVAL should be greater than 0");
      }
      dispatcherPoolSize = conf.getInt(
          YarnConfiguration.RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE,
          YarnConfiguration.
              DEFAULT_RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE);
      if (dispatcherPoolSize < 1) {
        throw new Exception("RM_SYSTEM_METRICS_PUBLISHER_DISPATCHER_POOL_SIZE should be greater than 0");
      }
      dispatcherBatchSize = conf.getInt(
          YarnConfiguration.RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE,
          YarnConfiguration.
              DEFAULT_RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE);
      if (dispatcherBatchSize < 2) {
        throw new Exception("RM_TIMELINE_SERVER_V1_PUBLISHER_DISPATCHER_BATCH_SIZE should be greater than 1");
      }
```